### PR TITLE
Refactor migrations to allow to easily add new test models

### LINKED ISSUE #6236: Refactor `MigrationStep.scala` to allow to easily add new test models
Right now, `MigrationStep.scala` is built as an ad-hoc application around the `Model.daml` test model. Factor out behaviors that can be re-used to test multiple models.

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
Closes #6236

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [x] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by stefanobaghino-da** [comment]
> Awesome, thanks a lot for cleaning up the mess I created! Very happy to get feedback from someone that knows Scala on my Scala code 

Not really a mess, I mostly broke pieces apart.

> One thing that is still missing and is just as important is the logic for validating the result from the step in the Haskell binary. This is currently tied completely to the specific model. I have some ideas for making that more generic but I’d definitely be interested in hearing your thoughts! (Let’s leave it out of this PR)

I know, but I opened a separate ticket for that (https://github.com/digital-asset/daml/issues/6237), I didn't want to do too much in a single PR.

File: compatibility/sandbox-migration/src/com/daml/JsonProtocol.scala [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  }
+
+  private implicit class JsObjectWith(val jsObject: JsObject) extends AnyVal {
+    def +(pair: (String, JsValue)): JsObject =
+      jsObject.copy(fields = jsObject.fields + pair)
+  }
+
+  import ProposeAccept._
+  implicit val ProposeAcceptCreatedResult: RootJsonFormat[CreatedResult] =
+    jsonFormat4(CreatedResult.apply)
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/src/com/daml/JsonProtocol.scala:28]: Yeah, I used the dirty trick of throwing exceptions to make library primitives cover this. :cry: 

I'm not very familiar with `spray-json`, it seems a bit limited in this regard (but probably I'm not simply able to use it properly).

File: compatibility/sandbox-migration/src/com/daml/Application.scala [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    def exercise(choice: Choice, contract: Contract, arguments: (String, Value)*): Future[Value] = {
+      val commands = Commands(
+        party = name,
+        commands = List(
+          Command().withExercise(
+            ExerciseCommand(
+              Some(choice.template.identifier),
+              contract.identifier,
+              choice.name,
+              Some(arguments.asRecordValue),
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply compatibility/sandbox-migration/src/com/daml/Application.scala:77]: Yeah, there are many arbitrary decisions here, I mostly wanted to split the concerns, I'm sure there are (and will be) multiple possible improvements.

File: compatibility/BUILD [hunk 3]
```diff
@@ -115,3 +115,15 @@ migration_test(
     tags = ["exclusive"],
     versions = platform_versions,
 ) if not is_windows else None
+
+migration_test(
+    name = "migration-test",
```
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply compatibility/BUILD:120]: I'll remove it for now and re-introduce it later. I was thinking that if you had a single test that covers all migrations in one go you can lower the chances of missing weird regressions when using this rule to quickly test while developing. Let me know what you think about it (and what name could make more sense: `migration-meta-test`, `meta-migration-test`, `test-migration-test`, `migration-test-test`? Something else?)

[the remark this answers is not in the document]
Comment (cocreature) [reply compatibility/BUILD]: Maybe just make it very explicit, `migration-1.0.0-to-head`? I don’t mind keeping it, I just find the name to be a bit too vague.

**Pull request merged**